### DISCUSSION TITLE
Import element zapper's scroll-unlocking heuristics into element picker

### DISCUSSION THREAD

**Discussion by partingscientist** [uBlockOrigin/uBlock-issues#2439] [Ideas] [answered]
Per the wiki,

> Additionally, depending on what element you are zapping, the element zapper will try to detect and deal with the web pages preventing scrolling through the entire content.

Currently, this behaviour is not present when using the element picker. I believe importing such behaviour would increase the user-friendliness of the picker, especially for those who understand using the element picker but not proficient enough in filter-crafting. Additionally, result(s) of the heuristic shall be added to the `My Filters` list in filter form, should this suggestion is accepted.

Here's a use case.
1. Navigate to [this article](https://theathletic.com/3085749/2022/01/27/premier-league-how-do-you-value-a-player/).
2. You will notice upon scrolling that a subscription offer will appear.
3. Reverting the page into a preferred behaviour is relatively trivial using the element zapper, only needing zaps for three elements.
4. However, doing so with the element picker is not trivial, for picking `###darken-overlay` does not unlock the page's body unlike with the zapper.
5. A preferred behaviour for the picker would be that upon clicking `Pick`, in addition with the current behaviour,
   - the page body's scrolling behaviour is restored, if needed, and
   - a filter to do so is added to `My Filters` alongside the element filter.

Thank you for considering.

**Comment by gothic-bum** [comment]
https://www.reddit.com/r/uBlockOrigin/comments/qpojmt/comment/hxp1lna/

> Zapper removes an element you select directly. Picker creates a filter to be able find the elements to hide when a page loads. It's easy to remove an element. The difficult part is to derive a filter from it so that it can be targeted automatically after a reload. That's what the picker does, *assist* you in creating filters.

**Reply by Yuki2718** [reply]
https://github.com/uBlockOrigin/uBlock-issues/issues/1058#issuecomment-631459847

**Comment by Yuki2718** [comment]
The suggestion is rather problematic to filter author who is the first and foremost user of the tool as we often have to preview :style() rules.

**Comment by uBlock-user** [comment]
Picker and Zapper exist separately for a reason. Asking to merge Zapper into Picker is not going to happen, this has been suggested before too and declined.

As for scroll unlocking feature, it already [exists](https://github.com/gorhill/uBlock/blob/8d9068c84ae32cc21efde3765a5c923e87cef377/src/js/scriptlets/epicker.js#L1000) in uBO too, but works sometimes and sometimes doesn't. That should be tweaked if possible.

**Comment by gorhill** [comment, marked as answer]
The way to scroll-unlock in the zapper can't be imported in the filter creation tool, it's done in a completely different manner. The zapper makes guesses as to whether scrolling is locked and apply changes in the hope this will restore scrolling -- not guaranteed to work (people should report when it does not work so that we can look into improving the "guesser").

With the picker, there is no guess, filter authors will create a very targeted filter which will disable scroll-lock.

The proper course of action in your example is to report the case to filter author so that it get properly fixed in an optimal manner in _Annoyances_ lists.

---

In the provided case, the filters would be:

    theathletic.com###slideup-paywall
    theathletic.com###darken-overlay
    theathletic.com##body.noscroll:style(position:unset!important;overflow:unset!important)

No heuristics in the zapper can come up with these filters, even the picker is unable to tell for sure what exact filter has to be created, there need to be a human trying out the various suggestions and often manually fine tuning using the browser dev tools to find the best one.

**Reply by krystian3w** [reply]
But not all anti annoying elements lists cover fake paywalls.
